### PR TITLE
[interp] Non-recursive newobj.

### PR DESCRIPTION
**Pull request by jaykrell**
A few wierd here, esp. for string, but CI passes.
Acknowledged but skipped for now, a request to also refactor the opcode -- it doesn't seem like much throwaway work to go in either order, even if the code is dead (which I don't know).

Also amend recent nonrecursive PR to simulate stack +2 instead of +1.

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build failed

File: mono/mini/interp/interp.c [hunk 1]
```diff
@@ -1944,7 +1944,7 @@ interp_entry (InterpEntryData *data)
 	ThreadContext *context;
 	stackval result;
 	stackval *args;
-	stackval *retval;
+	stackval *retval = &result;
```
Comment (jaykrell) [inline mono/mini/interp/interp.c]: This variable can just go away, and be replaced by `result` throughout the function.

Comment (jaykrell) [reply mono/mini/interp/interp.c]: Will do in a separate PR and undo here.

Comment (jaykrell) [reply mono/mini/interp/interp.c]: https://github.com/mono/mono/pull/18951

**Comment by jaykrell** [comment]
Last change was cosmetic. CI had already passed all lanes or at least all interpreter lanes.

File: mono/mini/interp/transform.c [hunk 2]
```diff
@@ -4562,6 +4562,8 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
 						}
 					}
 				} else {
+					// Runtime (interp_exec_method_full in interp.c) inserts extra stack to hold return value, before call.
+					simulate_runtime_stack_increase (td, 1);
```
Comment (BrzVlad) [inline mono/mini/interp/transform.c]: Hmm. Shouldn't this be +2 ? As well in the PR from the other day for non recursive VT ?

Comment (jaykrell) [reply mono/mini/interp/transform.c]: Should it have been +1 before?
I think yes to both.
“This” is .ctor signature or not, I meant to check.

Comment (jaykrell) [reply mono/mini/interp/transform.c]: I changed them both to 2.

**Comment by jaykrell** [comment]
While we prefer precision this does feel like a “better safe than sorry” situation. The cost of one higher than needed is tiny or even zero, the cost of one less than needed is memory corruption I presume (or zero).

**Comment by jaykrell** [comment]
https://github.com/mono/mono/issues/18817

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
I forgot to point out, it was a little wonky,
before the call, there are four cases, but after the call only three.
I guess it is equivalent, that `*sp = valuetype_this` covers
both `valuetype` cases, either `valuetype.p = vt_sp`, for entire `valuetype`.

File: mono/mini/interp/interp.c [hunk 3]
```diff
@@ -5032,60 +5025,50 @@ call:;
...
-				memset (&valuetype_this, 0, sizeof (stackval));
 				if (!m_class_is_enumtype (newobj_class) && (t->type == MONO_TYPE_VALUETYPE || (t->type == MONO_TYPE_GENERICINST && mono_type_generic_inst_is_valuetype (t)))) {
-					sp->data.p = vt_sp;
-					valuetype_this.data.p = vt_sp;
+					sp [0].data.p = vt_sp; // return value
+					sp [1].data.p = vt_sp; // first parameter
 				} else {
-					sp->data.p = &valuetype_this;
+					memset (sp, 0, sizeof (*sp));
+					sp [1].data.p = &sp [0].data; // first parameter is return value
```
Comment (jaykrell) [inline mono/mini/interp/interp.c]: If stackval were to have extra data at the start, it is not clear if this should be `&sp [0].data]` or `&sp [0]`.
It is presently moot though.

**Pull request merged**